### ISSUE TITLE
[MNG-7189] Remove undocumented environment variable MAVEN_BASEDIR in startup scripts

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8264]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** opened **[MNG-7189](https://issues.apache.org/jira/browse/MNG-7189?redirect=false)** and commented

During a review I have found this:

```
MAVEN_PROJECTBASEDIR="${MAVEN_BASEDIR:-`find_maven_basedir "$@"`}"
```

and

```
set "MAVEN_PROJECTBASEDIR=%MAVEN_BASEDIR%"
if not "%MAVEN_PROJECTBASEDIR%"=="" goto endDetectBaseDir
```

I was about to remove it until I have found:

```
$ grep -r MAVEN_BASEDIR .
./core-it-suite/src/test/java/org/apache/maven/it/MavenITmng5937MavenWrapper.java:        envVars.put( "MAVEN_BASEDIR", testDir.getAbsolutePath() );
./core-it-suite/src/test/java/org/apache/maven/it/MavenITmng5937MavenWrapper.java:        envVars.put( "MAVEN_BASEDIR", testDir.getAbsolutePath() );
./core-it-suite/src/test/java/org/apache/maven/it/MavenITmng5937MavenWrapper.java:        envVars.put( "MAVEN_BASEDIR", testDir.getAbsolutePath() );
./core-it-suite/src/test/java/org/apache/maven/it/MavenITmng5937MavenWrapper.java:        envVars.put( "MAVEN_BASEDIR", testDir.getAbsolutePath() );
./core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6118SubmoduleInvocation.java:        // It seems MAVEN_BASEDIR isn't always properly set, so make sure to have the right value here
./core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6118SubmoduleInvocation.java:        envVars.put( "MAVEN_BASEDIR", testDir.getAbsolutePath() );
./environments/solaris-10-jdk7/apache-maven-3.3.9.patch:< MAVEN_PROJECTBASEDIR="${MAVEN_BASEDIR:-$(find_maven_basedir)}"
./environments/solaris-10-jdk7/apache-maven-3.3.9.patch:> MAVEN_PROJECTBASEDIR="${MAVEN_BASEDIR:-`find_maven_basedir`}"
./environments/solaris-10-jdk8/apache-maven-3.3.9.patch:< MAVEN_PROJECTBASEDIR="${MAVEN_BASEDIR:-$(find_maven_basedir)}"
./environments/solaris-10-jdk8/apache-maven-3.3.9.patch:> MAVEN_PROJECTBASEDIR="${MAVEN_BASEDIR:-`find_maven_basedir`}"
```

It is not clear why this variable is there and and why it is necessary. There are several ways to resolve this issue:

* Remove it and rely on basedir discovery in scripts
* Document it officially in the prolog of the scripts
* Inline document them as private and used for testing purposes only

---

**Affects:** 3.0-alpha-1

**Remote Links:**
- [GitHub Pull Request #109
](https://github.com/apache/maven-integration-testing/pull/109)
- [GitHub Pull Request #490
](https://github.com/apache/maven/pull/490)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1)

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Maarten Mulders](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mthmulders) IIRC you made a change so we don't need to set this in ITs anymore, right?

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) the reason is that these tests are based on the `.mvn` directory, but the ITs picked up the wrong one, hence the ITs needed a hint to explicitly say where the maven-basedir for these tests is located.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

A bit confusing, if the IT needs a hint this means either the IT is incomplete or our discovery is, no?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[8ed9a1caa8890773b45c6c408a4e40acf4f4b0fd](https://github.com/apache/maven/commit/8ed9a1caa8890773b45c6c408a4e40acf4f4b0fd) introduced it.
It is there so you can execute maven from a deeper directory while picking up the .mvn configuration.
The .mvn is also important for the root-reactor aware feature in Maven 4.

With that I don't know what you want to be changed to the scripts (the ITs can be improved if my assumption is right).

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

My review is not necessary related to this issue, but noticed this discrepancy. So is `MAVEN_CONFIG` not documented and not clear why it is necessary/used.
I will create seperate issues for all other findings.

**Comment by jira-importer** [comment]
**[Maarten Mulders](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mthmulders)** commented

> Maarten Mulders IIRC you made a change so we don't need to set this in ITs anymore, right?

That's right, [2dc4f0cd](https://github.com/apache/maven-integration-testing/commit/2dc4f0cd187f4f464e94f183117e9797f634e99a) makes sure the "embedded launcher" that we use a lot in integration tests uses `maven.multiModuleProjectDirectory`.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Maarten Mulders](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mthmulders), yes, but this does not explain the env var and its usage. Do you think it is necessary? It is somewhat similar to the `GIT_DIR`.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

The comment says:

> 1. traverses directory structure from process work directory to filesystem root
> 2. first directory with .mvn subdirectory is considered project base directory

So it is not the same as execution root, most time it'll be next to the `.git` directory, however for Maven the `.mvn` is the marker as we shouldn't be bound to scm specific mechanisms, and I've seen projects where the Maven code is a subdirectory of the git repository, so another reason we cannot simply borrow `.git`

And yes we need to know the reactor root for at least MNG-6118.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I think there is a misunderstanding here. I am not about `MAVEN_PROJECTBASEDIR` or `maven.multiModuleProjectDirectory`, but solely about `MAVEN_BASEDIR`. This is the variable in question, undocumented.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

You're right, but it doesn't change the statement: It was introduced in [8ed9a1caa8890773b45c6c408a4e40acf4f4b0fd](https://github.com/apache/maven/commit/8ed9a1caa8890773b45c6c408a4e40acf4f4b0fd)
For me it is just guessing: MAVEN_BASEDIR is a way to override MAVEN_PROJECTBASEDIR, maybe for IDEs to benefit from, hence it made sense to use that for the integration tests.
We could try to remove it for Maven 4.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Removed, works locally. Created two PRs. Please test and review.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [b8cdac5573a27e97540dc1fd7b50fbe4de7d4410](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=b8cdac5573a27e97540dc1fd7b50fbe4de7d4410) and [1adb6bcebcee4c848b65f1956031a0ce1b172d32](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git;a=commit;h=1adb6bcebcee4c848b65f1956031a0ce1b172d32).
